core/src/main/java/hudson/model/Job.java

```diff
@@ -424,15 +427,24 @@ public synchronized void updateNextBuildNumber(int next) throws IOException {
     }
 
     /**
-     * Returns the configured build discarder for this job, or null if none.
+     * Returns the configured build discarder for this job, via {@link BuildDiscarderProperty}, or null if none.
      */
     public BuildDiscarder getBuildDiscarder() {
-        return logRotator;
+        BuildDiscarderProperty prop = getProperty(BuildDiscarderProperty.class);
```

Comment:
No, because in <code> setBuildDiscarder()</code> you remove the property and then set the new one. Without synchronization this method can return invalid null if it gets invoked in parallel